src/Silverback.Integration/Messaging/Encryption/DecryptorConsumerBehavior.cs

```diff
@@ -38,9 +38,22 @@ public DecryptorConsumerBehavior(ISilverbackCryptoStreamFactory streamFactory)
             Check.NotNull(next, nameof(next));
 
             if (context.Envelope.Endpoint.Encryption != null &&
-                !context.Envelope.Headers.Contains(DefaultMessageHeaders.Decrypted) &&
+                !context.Envelope.Headers.Contains(DefaultMessageHeaders.Decrypted) && // TODO: Sergio when/where is it set?
                 context.Envelope.RawMessage != null)
             {
+                if (context.Envelope.Headers.Contains(DefaultMessageHeaders.EncryptionKeyId) &&
```

Comment:
In my opinion it shouldn't matter if the key header was sent, if a callback/provider function is specified, it has to be used.
I also expect either a function or a fixed key, not both.